### PR TITLE
[arcilator] Add JIT runtime environment library and stdio hooks 

### PR DESCRIPTION
**Pull request by fzi-hielscher**
This PR adds a dynamically linked runtime environment library to arcilator, which makes functions declared in the `arcilator-runtime.h` available for JIT compiled runs.

The runtime is intended to provide support functions that can be called directly from the generated IR, similar to the `CRunnerUtils` in LLVM. I went for a dynamic instead of a static library because I think there is some (limited) use in having the option to swap the runtime environment without having to recompile arcilator itself.

Notably, it provides a well defined interface for certain methods of the C standard library (currently `fprintf`, `fputs`,  `fputc`). While it is technically possible to invoke them directly from the IR, going through the standard headers and a C preprocessor should avoid trouble caused by different libc implementations (glibc, musl, msvcrt, ucrt, ...).

PS: Just noticed that I switched from camel case to snake case. `_arcLibcFprintf` doesn't look right to me...

### CODE REVIEW

File: tools/arcilator/arcilator.cpp [hunk 1]
```diff
@@ -444,6 +450,8 @@ static LogicalResult processBuffer(
             llvm::errs() << "failed to create execution engine: "
                          << info.message() << "\n";
           });
+
+      arc_jit_runtime_env_deinit();
```
Comment (uenoku) [inline tools/arcilator/arcilator.cpp:454]: nit: you can use` llvm::make_scope_exit` to automatically call `deinit`.

File: tools/arcilator/arcilator-runtime.h [hunk 2]
```diff
@@ -1,12 +1,63 @@
...
+#error Unsupported endianess
+#endif
+#endif
+static_assert(sizeof(int) == 4, "Unsupported ABI");
+static_assert(sizeof(long long) == 8, "Unsupported ABI");
+
+// ---  Exports to the IR ---
+
+#ifdef _WIN32
+#define ARCEXPORT(rtype) extern "C" __declspec(dllexport) rtype __cdecl
```
Comment (uenoku) [inline tools/arcilator/arcilator-runtime.h:24]: Can you make this consistent with `MLIR_RUNNERUTILS_EXPORT` macro?

File: tools/arcilator/arcilator.cpp (hunk 1 above)
Comment (fzi-hielscher) [reply tools/arcilator/arcilator.cpp]: Nice. Thanks!

File: tools/arcilator/arcilator-runtime.h (hunk 2 above)
Comment (fzi-hielscher) [reply tools/arcilator/arcilator-runtime.h:24]: `MLIR_RUNNERUTILS_EXPORT` does not include `extern "C"` yet it looks like it is prefixed with `extern "C"` every time it is used. That seems odd to me.
The `__cdecl` is probably not required. It don't have access to my Windows PC right now, I'll check later today. Basically, I've just used every magic incantation here that I could come up with of to increase the likelihood of it linking. 😅

File: integration_test/arcilator/JIT/runtime-environment.mlir [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  llvm.mlir.global internal constant @global_hello("He%c%co \00") {addr_space = 0 : i32}
+  llvm.mlir.global internal constant @global_world("World\00") {addr_space = 0 : i32}
+
+  arc.model @dut io !hw.modty<input clk : !seq.clock> {
+  ^bb0(%arg0: !arc.storage<1>):
+    %cst0 = llvm.mlir.constant(0 : i32) : i32
+    %ascii_em = llvm.mlir.constant(33 : i32) : i32
+    %ascii_lf = llvm.mlir.constant(10 : i32) : i32
+    %ascii_l = llvm.mlir.constant(108 : i32) : i32
+    %in_clk = arc.root_input "clk", %arg0 {offset = 0 : i32} : (!arc.storage<1>) -> !arc.state<i1>
```
[the remark this answers is not in the document]
Comment (fzi-hielscher) [reply integration_test/arcilator/JIT/runtime-environment.mlir]: No, tbh I didn't even try without at least one port as I was unsure whether this is a supported use-case.
Looks like it is working just fine. 👍

File: tools/arcilator/arcilator-runtime.h [hunk 4]
```diff
@@ -1,12 +1,63 @@
...
+// Runtime Environment calls
+
+#define ARC_ENV_DECL_GET_PRINT_STREAM(idarg)                                   \
+  ARCEXPORT(FILE *) _arc_env_get_print_stream(uint32_t idarg)
+
+#ifndef ARC_NO_DEFAULT_GET_PRINT_STREAM
+ARC_ENV_DECL_GET_PRINT_STREAM(id) {
+  (void)id;
+  return stderr;
+}
```
[the remark this answers is not in the document]
Comment (fzi-hielscher) [reply tools/arcilator/arcilator-runtime.h:56]: Yeah, I've added this in anticipation of #7111. The problem is, AFAIK we don't have any lowering path that would support this, let alone an agreed on concept for file descriptors / output streams in the IR. I had a short exchange with @seldridge on this in the discord channel some weeks ago. For now we could at best add a command line flag to control this globally.

File: tools/arcilator/jit-env/arcilator-jit-env.h [hunk 5]
```diff
@@ -0,0 +1,23 @@
+// Arcilator JIT runtime environment API facing the arcilator.cpp
```
[the remark this answers is not in the document]
Comment (fzi-hielscher) [reply tools/arcilator/jit-env/arcilator-jit-env.h]: Yes, nice catch. I'm a little surprised the CI did not complain about it. The `arcilator-runtime.h` and `arcilator-header-cpp.py` should probably also at least get a license identifier, shouldn't they?

File: tools/arcilator/arcilator-runtime.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (maerhart) [reply tools/arcilator/arcilator-runtime.h:56]: Ah I see. I was just thinking about adding an if-statement here to allow the LLVM call operation to choose between stderr and stdout when writing MLIR integration tests and examples by hand. But this is not important at all. We can also just leave it as is until we have a proper story for file descriptors.

**Review by maerhart — APPROVED** [review]
LGTM! Thanks!

**Review by fabianschuiki — APPROVED** [review]
LGTM! Thanks for adding this 🥳 

Can't wait for us to finally get the VCD writing out of the runtime header and built into the model a bit more thoroughly. That will make the JIT pretty much equivalent to a separately-compiled model 😄

**Comment by fzi-hielscher** [comment]
That does make me wonder how to proceed from here. When writing this PR's code I didn't want to mess up existing build flows. So, the "arcilator-runtime" remains a header library and we simply provide a compiled binary of it to the JIT runner. That comes at the cost of mixing declarations and definitions in the header.

If our goal is to move stuff out of the header, maybe it would be better if we had an `arcilator-runtime.cpp` to put static (i.e., model independent) logic in? The compiled library would then be usable for both JIT and conventional flows. I'm specifically thinking of the parsing logic for `.mem` files. A header file is not a great place to put that. But I think dynamically building it into the MLIR model makes even less sense.

What do you think?

**Comment by fabianschuiki** [comment]
Hmmm yeah I think it would make sense to actually compile a runtime library, link against it from the JIT, and also install it as a standalone library alongside the rest of CIRCT. If you use arcilator to build an object file, you can link against that and optionally the runtime in whatever build flow you're using. And if you use arcilator in JIT mode, you also have the library available.

**Comment by fzi-hielscher** [comment]
I've brought the export macro closer to `MLIR_RUNNERUTILS_EXPORT` as @uenoku suggested. On my local runs the test produces the expected output for Windows, MacOS, and Linux (glibc). So, I'm about 50% sure this will not break in the nightly tests. 😬

**Comment by uenoku** [comment]
Thank you @fzi-hielscher!

**Pull request merged**